src/coprocessor/dag/batch_executor/scan_executor.rs

```diff
@@ -0,0 +1,228 @@
+                    self.point_get(r)?
+                }
+                super::ranges_iter::IterStatus::NewNonPointRange(r) => {
+                    self.scanned_rows_per_range.push(0);
+                    self.reset_range(r)?;
+                    self.scan_next()?
+                }
+                super::ranges_iter::IterStatus::Continue => self.scan_next()?,
+                super::ranges_iter::IterStatus::Drained => {
+                    return Ok(true); // drained
+                }
+            };
+            if let Some((key, value)) = some_row {
+                // Retrieved one row from point range or non-point range.
+                self.scanned_rows_per_range
+                    .last_mut()
+                    .map_or((), |val| *val += 1);
+
+                self.imp.process_kv_pair(&key, &value, columns)?;
+
+                if columns.rows_len() >= expect_rows {
+                    return Ok(false); // not drained
+                }
+            } else {
+                // No more row in the range.
+                self.ranges.notify_drained();
+            }
+        }
+    }
+}
+
+/// Extracts `FieldType` from `ColumnInfo`.
+// TODO: Embed FieldType in ColumnInfo directly in Cop DAG v2 to remove this function.
+pub fn field_type_from_column_info(ci: &ColumnInfo) -> FieldType {
+    let mut field_type = FieldType::new();
+    field_type.set_tp(ci.get_tp());
+    field_type.set_flag(ci.get_flag() as u32); // FIXME: This `as u32` is really awful.
+    field_type.set_flen(ci.get_columnLen());
+    field_type.set_decimal(ci.get_decimal());
+    field_type.set_collate(ci.get_collation());
+    // Note: Charset is not provided in column info.
+    field_type
+}
+
+impl<S: Store, I: ScanExecutorImpl, P: PointRangePolicy> BatchExecutor for ScanExecutor<S, I, P> {
+    #[inline]
+    fn schema(&self) -> &[FieldType] {
+        self.imp.schema()
+    }
+
+    #[inline]
+    fn next_batch(&mut self, expect_rows: usize) -> BatchExecuteResult {
+        assert!(!self.is_ended);
+        assert!(expect_rows > 0);
+
+        let mut data = self.imp.build_column_vec(expect_rows);
+        let is_drained = self.fill_column_vec(expect_rows, &mut data);
+
+        // After calling `fill_column_vec`, columns' length may not be identical when some of the
+        // columns are correctly decoded while others are not. So let's truncate columns.
+        data.truncate_into_equal_length();
+
+        // TODO
+        // If `is_drained.is_err()`, it means that there is an error after *successfully* retrieving
+        // these rows. After that, if we only consumes some of the rows (TopN / Limit), we should
+        // ignore this error.
+
+        match &is_drained {
+            Err(_) => self.is_ended = true,
+            Ok(true) => self.is_ended = true,
+            Ok(false) => {}
+        };
+
+        BatchExecuteResult {
+            data,
+            is_drained,
+            warnings: self.imp.mut_context().take_warnings(),
```

Comment:
I think we should not take warnings in each `next_batch`, just take them when it is drained because:
if the number of warnings is more than the max count, we needn't collect more.

Reply:
We need row number for each batch in order to correctly map warnings into rows (and then ignore these in unused rows) in future. Thus, we need to return warnings batch by batch instead of waiting for them being accumulated (which means calculating its accurate row will be more complicated).